### ISSUE TITLE
 How can I train TransfoXLLMHeadModel on a dataset that is different than just WikiText103?

### ISSUE THREAD

**Issue by h56cho** [pytorch/text#630]
Hello,

Is it possible to train HuggingFace TransfoXLLMHeadModel on a dataset different than WikiText103, say, on the combined WikiText2 and WikiText103 dataset?

Below are my code:
```js

# Import packages 
import torch
import torch.nn as nn
import torch.nn.functional as F
from transformers import TransfoXLConfig, TransfoXLTokenizer, TransfoXLLMHeadModel
from transformers import AdamW, WarmupLinearSchedule
import spacy
import torchtext
from torchtext.data.utils import get_tokenizer
from torchtext.data import Field, BPTTIterator, TabularDataset 
import tensorflow as tf
import math
import random
import numpy as np
import pandas as pd 
import time

# set hyperparameters for this experiment
bptt = 30
batch_size = 64
lr = 0.01 # learning rate
criterion = nn.CrossEntropyLoss() # loss criterion

# define tokenizer
en = spacy.load('en')

def Sp_Tokenizer(text): 
    return [tok.text for tok in en.tokenizer(text)]

# define the English text field
TEXT = Field(tokenize = Sp_Tokenizer,
             init_token='< sos >',
             eos_token='< eos >',
             unk_token='< unk >',
             tokenizer_language='en',
             lower=True)

# load the datasets
train_Wiki2, val_Wiki2, test_Wiki2 = torchtext.datasets.WikiText2.splits(TEXT)
train_Wiki103, val_Wiki103, test_Wiki103 = torchtext.datasets.WikiText103.splits(TEXT)

# Define device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# build vocabulary based on the defined field and
# the combined WikiText2 and WikiText103 datasets.
TEXT.build_vocab(train_Wiki2, val_Wiki2, test_Wiki2, 
                 train_Wiki103, val_Wiki103, test_Wiki103)

# set hyperparameter ntokens
ntokens = len(TEXT.vocab.stoi)

## specify the transformer-XL model that we are going to use.
#
# define transformer-XL configuration.
transfoXLconfig = TransfoXLConfig(vocab_size_or_config_json_file = ntokens,
                                  cutoffs = [20000, 40000, 200000], 
                                  d_model = 64, 
                                  d_embed = 64, 
                                  n_head = 16, 
                                  d_head = 64,
                                  n_layer = 5,
                                  attn_type = 0,
                                  dropout = 0.1, 
                                  output_hidden_states = True,
                                  output_attentions = True)

# define the transformer-XL model based on the specified configuration.
model = TransfoXLLMHeadModel(transfoXLconfig)

# add new tokens to the embeddings of our model
model.resize_token_embeddings(ntokens)

# define BPTTiterators
#
train_iter, val_iter, test_iter = BPTTIterator.splits(
            (train_Wiki2, val_Wiki2, test_Wiki2),
            batch_size = batch_size,
            bptt_len= bptt,
            sort_key=lambda x: len(x.text),
            sort_within_batch = True,
            shuffle = False,
            device= device,
            repeat=False)

train = next(iter(train_iter))
val = next(iter(train_iter))
test = next(iter(test_iter))

```

and now I am trying to write the train function but I am not sure how exactly I should proceed.
Below is what I tried:

```js
# define the hyperparameters for running the train function.
train = train 
optimizer = AdamW(model.parameters())
scheduler = WarmupLinearSchedule(optimizer = optimizer, 
                                 warmup_steps = 200, 
                                 t_total = 1000,
                                 last_epoch = -1)

model.train()



# define the train function
def train(model, train, bptt, criterion, optimizer, scheduler, ntokens, log_interval):

    # initialize total_loss to 0
    total_loss = 0
    # measure the computation time
    start_time = time.time()
    # number of tokens in the vocabulary
    ntokens = ntokens 

    for i in range(train.text.size()[1]):

        batch = i
        input_ids, targets = train.text[:,i], train.target[:,i]
        input_ids = torch.tensor(input_ids.tolist()).unsqueeze(0) 
        optimizer.zero_grad()
        
        output = model(input_ids)
        loss = criterion(output.view(-1, ntokens), targets) # not sure on this
        
        # calculate gradient by backwarding the loss
        # calculate gradient of the loss w.r.t weights
        loss.backward()
        
        # clips gradient norm of an iterable of parameters.
        # The norm is computed over all gradients together, as if they were
        # concatenated into a single vector. Gradients are modified in-place.
        # so basically just normalizes the weights and returns them.
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
        optimizer.step() # update the weights by following the WarmupLinearSchedule for the lr.

        # update the with the calculated loss 
        total_loss = total_loss + loss.item() 
        
        if batch % log_interval == 0 and batch > 0:
            cur_loss = total_loss / log_interval
            elapsed = time.time() - start_time
            print('| epoch {:3d} | {:5d}/{:5d} batches | '
                  'lr {:02.2f} | ms/batch {:5.2f} | '
                  'loss {:5.2f} | ppl {:8.2f}'.format(
                    epoch, batch, train.text.size()[1] // bptt, scheduler.get_lr()[0],
                    elapsed * 1000 / log_interval,
                    cur_loss, math.exp(cur_loss)))
            total_loss = 0
            start_time = time.time()
```

But I don't think the line `loss = criterion(output.view(-1, ntokens), targets)` shouldn't work since the line `output = model(input_ids)` does not actually give out the final output from the model, but it rather gives out (according to the HuggingFace documentation) prediction_scores, mems, attention, etc. How can I train TransfoXLLMHeadModel on a dataset that is different than just WikiText103?

Thank you,

**Comment by zhangguanheng66** [comment]
https://github.com/pytorch/text/issues/609

**Comment by h56cho** [comment]
Hello,

Thank you for your comment.
I already have seen the post that you have pointed out, and was trying to follow the instructions but I am still having troubles with training my custom Transformer-XL model on WikiText2 dataset. Below are my BPTTiterator and the train function.

Do you see any error in my train function? Thank you,

```js
train_iter, val_iter, test_iter = BPTTIterator.splits(
            (new_train_Wiki2, val_Wiki2, test_Wiki2),
            batch_size = batch_size,
            bptt_len= bptt,
            sort_key=lambda x: len(x.text),
            sort_within_batch = True,
            shuffle = False,
            device= device,
            repeat=False)

train = next(iter(train_iter))
val = next(iter(train_iter))
test = next(iter(test_iter))

model.train()


# define the train function
def train(model, train, bptt, criterion, optimizer, scheduler, ntokens, log_interval):

    # initialize total_loss to 0
    total_loss = 0
    # measure the computation time
    start_time = time.time()
    # number of tokens in the vocabulary
    ntokens = ntokens 

    for i in range(train.text.size()[1]):
        
        # import tensorflow as tf
        # from transformers import TransfoXLTokenizer, TFTransfoXLLMHeadModel
        # tokenizer2 = TransfoXLTokenizer.from_pretrained('transfo-xl-wt103')
        # model = TFTransfoXLLMHeadModel.from_pretrained('transfo-xl-wt103')
        # input_ids = torch.tensor(tokenizer.encode("Hello, my dog is cute")).unsqueeze(0)  # Batch size 1
        # outputs = model(input_ids)
        # prediction_scores, mems = outputs[:2]
        
        batch = i
        input_ids, targets = train.text[:,i], train.target[:,i]
        input_ids = torch.tensor(input_ids.tolist()).unsqueeze(0) 
        targets = torch.tensor(targets.tolist()).unsqueeze(0)
        optimizer.zero_grad() # set the gradient back to 0 (necessary step)
        output = model(input_ids, labels = targets)
        loss = output[0]

        loss.backward()
        
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
        optimizer.step()

**Comment by h56cho** [comment]
I am especially interested in knowing whethe the part below is correct:
```
output = model(input_ids, labels = targets)
loss = output[0]
loss.backward()
```
What I am intending to do here is to train my custom Transformer-XL model by calculating the loss and propagate it backwards. 

PS: It seems that this HuggingFace integration is not allowing users to specify the type of loss that's being calculated --- in the `loss = output[0]`, what type of loss does this specifically refer to? Thank you,
